syntax = "proto3";

option go_package = "github.com/hatchet-dev/hatchet/internal/services/dispatcher/contracts";

import "google/protobuf/timestamp.proto";

service Dispatcher {
    rpc Register(WorkerRegisterRequest) returns (WorkerRegisterResponse) {}

    rpc Listen(WorkerListenRequest) returns (stream AssignedAction) {}

    rpc SendActionEvent(ActionEvent) returns (ActionEventResponse) {}

    rpc Unsubscribe(WorkerUnsubscribeRequest) returns (WorkerUnsubscribeResponse) {}
}

message WorkerRegisterRequest {
    // the name of the worker
    string workerName = 1;

    // a list of actions that this worker can run
    repeated string actions = 2;

    // (optional) the services for this worker
    repeated string services = 3;
}

message WorkerRegisterResponse {
    // the tenant id
    string tenantId = 1;

    // the id of the worker
    string workerId = 2;

    // the name of the worker
    string workerName = 3;
}

enum ActionType {
    START_STEP_RUN = 0;
    CANCEL_STEP_RUN = 1;
}

message AssignedAction {
    // the tenant id
    string tenantId = 1;

    // the job id
    string jobId = 2;

    // the job name
    string jobName = 3;

    // the job run id
    string jobRunId = 4;

    // the step id
    string stepId = 5;

    // the step run id
    string stepRunId = 6;

    // the action id
    string actionId = 7;

    // the action type
    ActionType actionType = 8;

    // the action payload
    string actionPayload = 9;
}

message WorkerListenRequest {
    // the id of the worker
    string workerId = 1;
}

message WorkerUnsubscribeRequest {
    // the id of the worker
    string workerId = 1;
}

message WorkerUnsubscribeResponse {
    // the tenant id to unsubscribe from
    string tenantId = 1;

    // the id of the worker
    string workerId = 2;
}

enum ActionEventType {
    STEP_EVENT_TYPE_UNKNOWN = 0;
    STEP_EVENT_TYPE_STARTED = 1;
    STEP_EVENT_TYPE_COMPLETED = 2;
    STEP_EVENT_TYPE_FAILED = 3;
}

message ActionEvent {
    // the id of the worker
    string workerId = 1;

    // the id of the job
    string jobId = 2;

    // the job run id
    string jobRunId = 3;

    // the id of the step
    string stepId = 4;

    // the step run id
    string stepRunId = 5;

    // the action id
    string actionId = 6;

    google.protobuf.Timestamp eventTimestamp = 7;

    // the step event type
    ActionEventType eventType = 8;

    // the event payload
    string eventPayload = 9;
}

message ActionEventResponse {
    // the tenant id
    string tenantId = 1;

    // the id of the worker
    string workerId = 2;
}
